modules/affile/logrotate.c

```diff
@@ -0,0 +1,174 @@
+}
+
+gchar *get_log_file_name(const gchar *filename, gsize rotation_suffix)
+{
+  if (rotation_suffix == 0)
+    {
+      return g_strdup_printf("%s", filename);
+    }
+  else
+    {
+      return g_strdup_printf("%s.%ld", filename, rotation_suffix);
+    }
+}
+
+gsize first_log_file_to_rotate(const gchar *filename, gsize max_rotations)
+{
+  gchar *log_filename;
+  for (gsize i = 1; i <= max_rotations; i++)
+    {
+      log_filename = get_log_file_name(filename, i);
+      gboolean log_file_exists = g_file_test(log_filename, G_FILE_TEST_EXISTS);
+      g_free(log_filename);
+
+      if (!log_file_exists)
+        {
+          return i-1;
+        }
+    }
+
+  return max_rotations;
+}
+
+gint rotate_log_file(const gchar *filename, gsize roation_suffix, gsize max_rotations)
+{
+  gchar *current_filename = get_log_file_name(filename, roation_suffix);
+  gint res = 0;
+
+  if (g_file_test(current_filename, G_FILE_TEST_EXISTS))
+    {
+      if (roation_suffix == max_rotations)
+        {
+          msg_debug("Deleting oldest log file",
+                    evt_tag_str("filename", current_filename));
+          res = remove(current_filename);
+        }
+      else
+        {
+          gchar *rotated_filename = get_log_file_name(filename, roation_suffix+1);
+          msg_debug("Rotating log file",
+                    evt_tag_str("filename", current_filename),
+                    evt_tag_str("new_filename", rotated_filename));
+          res = rename(current_filename, rotated_filename);
+          g_free(rotated_filename);
+        }
+
+      if (res == -1)
+        {
+          msg_error("Error renaming or deleting log file during log-rotation",
+                    evt_tag_str("filename", current_filename),
+                    evt_tag_errno("errno", errno));
+        }
+    }
+  else
+    {
+      msg_error("Error could not rotate log file because it does not exist",
+                evt_tag_str("filename", current_filename),
+                evt_tag_errno("errno", errno));
+      res = -1;
+    }
+
+  g_free(current_filename);
+
+  return res;
+}
+
+/* TODO:
+ * Do renaming in separate thread?
+ * i.e. in writer thread delete oldest log file and create a new temp logfile --> filename_tmp
+ * the writer only does this if not already a tmp file is created (otherwise the renaming thread is working)
+ * checked is this condition via the gcond-lock waiting for the thread to finish.
+ *
+ * The renaming thread renames all other files including the tmp file to filename.1
+ *
+ */
+LogRotateStatus logrotate_do_rotate(LogRotateOptions *logrotate_options, const gchar *filename)
+{
+  if (logrotate_options == NULL || filename == NULL)
+    {
+      return LR_ERROR;
+    }
+
+  msg_info("Maximum log file size reached, rotating log file ... ",
+           evt_tag_str("filename", filename));
+
+  // (1) check if main log file exists
+  if (!g_file_test(filename, G_FILE_TEST_EXISTS))
+    {
+      msg_warning("Main log file does not exist",
```

Comment:
Please replace the log string by s.g like
"WARNING: Logrotate could not find the main log file"
as the warning lines have WARNING: prefix everywhere, and mentioning Logrotate can identify the sender better